### ISSUE TITLE
What is the reason for the particularly poor accuracy of the training model by training the encode feature extractor first and then the decode method

### ISSUE THREAD

**Issue by Liumeifeng283** [babahui/FPC#6]
Do you need any special Settings
this is trian.py
`from utils.logger import setup_logger
from datasets import make_dataloader
from model import make_model
from solver import make_optimizer
from solver.scheduler_factory import create_scheduler
from loss import make_loss
from processor import do_train,train_decode
import random
import torch
import numpy as np
import os
import argparse
import torch.nn as nn
# from timm.scheduler import create_scheduler
from config import cfg
## 一个两阶段式的训练，首先训练一个稀疏的encode模块，在训练decode模块
def load_model(pretrained_path, model):
    state_dict = torch.load(pretrained_path, map_location="cpu")
    out_dict = {}
    if 'model' in state_dict:
        # For deit models
        state_dict = state_dict['model']
    for k, v in state_dict.items():
        out_dict[k] = v
    return out_dict

def set_seed(seed):
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = True

def freeze_rec(model):
    model.base.requires_grad_(True)
    model.b1.requires_grad_(False)
    model.classifier.requires_grad_(False)
    model.bottleneck.requires_grad_(False)
    model.rec_b.requires_grad_(False)
    model.rec_classifier.requires_grad_(False)
    model.rec_bottleneck.requires_grad_(False)

def unfreeze_rec(model):
    model.base.requires_grad_(False)
    model.b1.requires_grad_(True)
    model.classifier.requires_grad_(True)
    model.bottleneck.requires_grad_(True)
    model.rec_b.requires_grad_(True)
    model.rec_classifier.requires_grad_(True)
    model.rec_bottleneck.requires_grad_(True)

if __name__ == '__main__':

    parser = argparse.ArgumentParser(description="ReID Baseline Training")
    parser.add_argument(
        "--config_file", default="", help="path to config file", type=str
    )

    parser.add_argument("opts", help="Modify config options using the command-line", default=None,
                        nargs=argparse.REMAINDER)
    parser.add_argument("--local_rank", default=0, type=int)
    args = parser.parse_args()

    if args.config_file != "":
        cfg.merge_from_file(args.config_file)
    cfg.merge_from_list(args.opts)
    set_seed(cfg.SOLVER.SEED)

    if cfg.MODEL.DIST_TRAIN:
        torch.cuda.set_device(args.local_rank)

    output_dir = cfg.OUTPUT_DIR
    if output_dir and not os.path.exists(output_dir):
        os.makedirs(output_dir)

    logger = setup_logger("fpc", output_dir, if_train=True)
    logger.info("Saving model in the path :{}".format(cfg.OUTPUT_DIR))
    logger.info(args)

    if args.config_file != "":
        logger.info("Loaded configuration file {}".format(args.config_file))
        with open(args.config_file, 'r') as cf:
            config_str = "\n" + cf.read()
            logger.info(config_str)
    logger.info("Running with config:\n{}".format(cfg))

    if cfg.MODEL.DIST_TRAIN:
        torch.distributed.init_process_group(backend='nccl', init_method='env://')

    os.environ['CUDA_VISIBLE_DEVICES'] = cfg.MODEL.DEVICE_ID
    train_loader, train_loader_normal, val_loader, num_query, num_classes, camera_num, view_num, query_loader, gallery_loader = make_dataloader(cfg)
    # 训练encode模块是no reb
    model = make_model(cfg, num_class=num_classes, camera_num=camera_num, view_num = view_num)
    loss_func, center_criterion = make_loss(cfg, num_classes=num_classes)

    optimizer, optimizer_center = make_optimizer(cfg, model, center_criterion)
    scheduler = create_scheduler(cfg, optimizer)

    # approach to reduce model inference time in memory constuction. For a detailed discussion, refer to the appendix section of the paper.
    device = "cuda"
    if device:
        if torch.cuda.device_count() > 1:
            print('Using {} GPUs for inference'.format(torch.cuda.device_count()))
            model = nn.DataParallel(model)
        model.to(device)
    model.train()
    freeze_rec(model)
    do_train(
        cfg,
        model,
        center_criterion,
        train_loader,
        optimizer,
        optimizer_center,
        scheduler,
        loss_func,
        num_query,
        local_rank=args.local_rank
    )
   # copy encode模块且将fixed
    unfreeze_rec(model)
    cfg.SOLVER.IMS_PER_BATCH = 16
    cfg.SOLVER.MAX_EPOCHS = 20
    train_loader, train_loader_normal, val_loader, num_query, num_classes, camera_num, view_num, query_loader, gallery_loader = make_dataloader(
        cfg)
    initialized = False
    start_idx = 0
    # 添加恢复模块的内容
    for n_iter, (img, pid, camid, camids, target_view, imgpath) in enumerate(train_loader):
        with torch.no_grad():
            img = img.to(device)
            camids = camids.to(device)
            target_view = target_view.to(device)
            target = pid.to(device)

            memory_g_cls, memory_p = model(img, target, cam_label=camids, view_label=target_view, get_memory=True,
                                           keep_rate=1.0)

            # Initialize tensors on the first iteration
            if not initialized:
                total_size = len(train_loader.dataset)
                dim1 = memory_g_cls.shape[1:]  # Get all dimensions except the first
                dim2 = memory_p.shape[1:]  # Get all dimensions except the first
                memory_global_cls = torch.zeros((total_size, *dim1))
                memory_patch = torch.zeros((total_size, *dim2))
                initialized = True

            end_idx = start_idx + img.size(0)
            memory_global_cls[start_idx:end_idx, ...] = memory_g_cls.cpu()
            memory_patch[start_idx:end_idx, ...] = memory_p.cpu()
            start_idx = end_idx

    # Split the data
    memory_global_cls = memory_global_cls[num_query:]
    gallery_memory_patch = memory_patch[num_query:]

    train_decode(
        cfg,
        model,
        center_criterion,
        train_loader,
        val_loader,
        optimizer,
        optimizer_center,
        scheduler,
        loss_func,
        num_query,
        memory_global_cls,
        gallery_memory_patch,
        args.local_rank
    )
`
this is proccess.py
`def do_train(cfg,
             model,
             center_criterion,
             train_loader,
             optimizer,
             optimizer_center,
             scheduler,
             loss_fn,
             num_query,
             local_rank):
    log_period = cfg.SOLVER.LOG_PERIOD
    checkpoint_period = cfg.SOLVER.CHECKPOINT_PERIOD
    eval_period = cfg.SOLVER.EVAL_PERIOD

    device = "cuda"
    epochs = cfg.SOLVER.MAX_EPOCHS

    logger = logging.getLogger("fpc.train")
    logger.info('start training')
    _LOCAL_PROCESS_GROUP = None
    if device:
        model.to(local_rank)
        if torch.cuda.device_count() > 1 and cfg.MODEL.DIST_TRAIN:
            print('Using {} GPUs for training'.format(torch.cuda.device_count()))
            model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[local_rank], find_unused_parameters=True)

    loss_meter = AverageMeter()
    acc_meter = AverageMeter()

    evaluator = R1_mAP_eval(num_query, max_rank=50, feat_norm=cfg.TEST.FEAT_NORM)
    scaler = amp.GradScaler()
    # train encode 部分
    for epoch in range(1, epochs + 1):
        start_time = time.time()
        loss_meter.reset()
        acc_meter.reset()
        evaluator.reset()
        scheduler.step(epoch)
        model.train()
        # (img, vid, target_cam, target_view)
        for n_iter, (img, pid, camid, camids, target_view, imgpath)  in enumerate(train_loader):
            optimizer.zero_grad()
            optimizer_center.zero_grad()
            img = img.to(device)
            camids = camids.to(device)
            target = pid.to(device)
            target_cam = camid.to(device)
            target_view = target_view.to(device)
            with amp.autocast(enabled=True):
                score,feat,_,_ = model(img, label=target, cam_label=camids, view_label=target_view, get_memory=False,
                              reconstruction=False, imgpath=imgpath, keep_rate=0.8)

                # Loss = ID loss + Tri Loss
                loss = loss_fn(score, feat, target, target_cam)

            scaler.scale(loss).backward()

            scaler.step(optimizer)
            scaler.update()

            if 'center' in cfg.MODEL.METRIC_LOSS_TYPE:
                for param in center_criterion.parameters():
                    param.grad.data *= (1. / cfg.SOLVER.CENTER_LOSS_WEIGHT)
                scaler.step(optimizer_center)
                scaler.update()
            if isinstance(score, list):
                acc = (score[0].max(1)[1] == target).float().mean()
            else:
                acc = (score.max(1)[1] == target).float().mean()

            loss_meter.update(loss.item(), img.shape[0])
            acc_meter.update(acc, 1)

        end_time = time.time()
        time_per_batch = (end_time - start_time) / (n_iter + 1)
        if cfg.MODEL.DIST_TRAIN:
            pass
        else:
            logger.info("Epoch[{}]  Loss: {:.3f}, Acc: {:.3f}, Base Lr: {:.2e}"
                        .format(epoch, loss_meter.avg, acc_meter.avg, scheduler._get_lr(epoch)[0]))

        if epoch % checkpoint_period == 0:
            if cfg.MODEL.DIST_TRAIN:
                if dist.get_rank() == 0:
                    torch.save(model.state_dict(),
                               os.path.join(cfg.OUTPUT_DIR, cfg.MODEL.NAME + '_{}.pth'.format(epoch)))
            else:
                torch.save(model.state_dict(),
                    os.path.join(cfg.OUTPUT_DIR, cfg.MODEL.NAME + '_{}.pth'.format(epoch)))

    # 训练decode模块

def train_decode(cfg,
             model,
             center_criterion,
             train_loader,
             val_loader,
             optimizer,
             optimizer_center,
             scheduler,
             loss_fn,
             num_query,
             memory_global_cls,
             gallery_memory_patch,
             local_rank):
    log_period = cfg.SOLVER.LOG_PERIOD
    checkpoint_period = cfg.SOLVER.CHECKPOINT_PERIOD
    eval_period = cfg.SOLVER.EVAL_PERIOD

    device = "cuda"
    epochs = cfg.SOLVER.MAX_EPOCHS

    logger = logging.getLogger("fpc.train")
    logger.info('start training')
    _LOCAL_PROCESS_GROUP = None
    if device:
        model.to(local_rank)
        if torch.cuda.device_count() > 1 and cfg.MODEL.DIST_TRAIN:
            print('Using {} GPUs for training'.format(torch.cuda.device_count()))
            model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[local_rank],
                                                              find_unused_parameters=True)

    loss_meter = AverageMeter()
    acc_meter = AverageMeter()

    evaluator = R1_mAP_eval(num_query, max_rank=50, feat_norm=cfg.TEST.FEAT_NORM)
    scaler = amp.GradScaler()
    # train encode 部分
    for epoch in range(1, epochs + 1):
        start_time = time.time()
        loss_meter.reset()
        acc_meter.reset()
        evaluator.reset()
        scheduler.step(epoch)
        model.train()
        # (img, vid, target_cam, target_view)
        for n_iter, (img, pid, camid, camids, target_view, imgpath) in enumerate(train_loader):
            optimizer.zero_grad()
            # optimizer_center.zero_grad()
            img = img.to(device)
            camids = camids.to(device)
            target = pid.to(device)
            target_cam = camid.to(device)
            target_view = target_view.to(device)
            memory_global_cls = memory_global_cls.to(device)
            gallery_memory_patch = gallery_memory_patch.to(device)
            with amp.autocast(enabled=True):
                score, feat, _, _ = model(img, label=target, cam_label=camids, view_label=target_view, get_memory=False,
                                          memory_global_cls=memory_global_cls,
                                          gallery_memory_patch=gallery_memory_patch,
                                          reconstruction=True, imgpath=imgpath,keep_rate=0.8)

                # Loss = ID loss + Tri Loss
                loss = loss_fn(score, feat, target, target_cam)

            scaler.scale(loss).backward()

            scaler.step(optimizer)
            scaler.update()

            if 'center' in cfg.MODEL.METRIC_LOSS_TYPE:
                for param in center_criterion.parameters():
                    param.grad.data *= (1. / cfg.SOLVER.CENTER_LOSS_WEIGHT)
                scaler.step(optimizer_center)
                scaler.update()
            if isinstance(score, list):
                acc = (score[0].max(1)[1] == target).float().mean()
            else:
                acc = (score.max(1)[1] == target).float().mean()

            loss_meter.update(loss.item(), img.shape[0])
            acc_meter.update(acc, 1)

        end_time = time.time()
        time_per_batch = (end_time - start_time) / (n_iter + 1)
        if cfg.MODEL.DIST_TRAIN:
            pass
        else:
            logger.info("Epoch[{}]  Loss: {:.3f}, Acc: {:.3f}, Base Lr: {:.2e}"
                        .format(epoch, loss_meter.avg, acc_meter.avg, scheduler._get_lr(epoch)[0]))

        if epoch % checkpoint_period == 0:
            if cfg.MODEL.DIST_TRAIN:
                if dist.get_rank() == 0:
                    torch.save(model.state_dict(),
                               os.path.join(cfg.OUTPUT_DIR, cfg.MODEL.NAME + '_decode_{}.pth'.format(epoch)))
            else:
                torch.save(model.state_dict(),
                           os.path.join(cfg.OUTPUT_DIR, cfg.MODEL.NAME + '_decode_{}.pth'.format(epoch)))

        if epoch % epochs == 0:
            if cfg.MODEL.DIST_TRAIN:
                pass
            else:
                model.eval()
                for n_iter, (img, vid, camid, camids, target_view, _) in enumerate(val_loader):
                    with torch.no_grad():
                        img = img.to(device)
                        camids = camids.to(device)
                        target_view = target_view.to(device)
                        feat, idx = model(img, label=target, cam_label=camids, view_label=target_view, get_memory=False,
                                          memory_global_cls=None, memory_local_cls=None,
                                          gallery_memory_patch=None,
                                          reconstruction=False, imgpath=imgpath, keep_rate=0.8)
                        evaluator.update((feat, pid, camid))
                cmc, mAP, _, _, _, _, _ = evaluator.compute()
                logger.info("Validation Results - Epoch: {}".format(epoch))
                logger.info("mAP: {:.1%}".format(mAP))
                for r in [1, 5, 10]:
                    logger.info("CMC curve, Rank-{:<3}:{:.1%}".format(r, cmc[r - 1]))
`

**Comment by Jessical7** [comment]
您好，请问是否方便交流一下FPC的两阶段训练呢，我的qq是274235013，邮箱是274235013@qq.com
